### PR TITLE
[Merged by Bors] - feat(Data/Finset/Card): the number of element satisfying `P` is no greater than `n` iff every subset with more than `n` elements contains an element not satisfying `P`

### PR DESCRIPTION
**Pull request by jjaassoonn**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Multiset/Basic.lean [hunk 1]
```diff
@@ -2154,6 +2154,19 @@ lemma map_filter' {f : α → β} (hf : Injective f) (s : Multiset α)
...
+  fconstructor
+  · intro H s' hs' s'_card
+    by_contra! rid
+    have card1 := card_le_of_le (monotone_filter_left P hs') |>.trans H
+    have card2 : card (s'.filter P) = card s'
+    · rw [filter_eq_self.mpr rid]
+    exact lt_irrefl _ <| lt_of_lt_of_le (card2.symm ▸ s'_card) card1
+  · contrapose!
+    intro H
+    exact ⟨s.filter P, filter_le _ _, H, fun a ha ↦ (mem_filter.mp ha).2⟩
```
Comment (alreadydone) [inline Mathlib/Data/Multiset/Basic.lean:2168]:
```suggestion
    have card := card_le_of_le (monotone_filter_left P hs') |>.trans H
    exact s'_card.not_le (filter_eq_self.mpr rid ▸ card)
  · contrapose!
    exact fun H ↦ ⟨s.filter P, filter_le _ _, H, fun a ha ↦ (mem_filter.mp ha).2⟩
```

File: Mathlib/Data/Finset/Card.lean [hunk 2]
```diff
@@ -267,6 +267,11 @@ theorem fiber_card_ne_zero_iff_mem_image (s : Finset α) (f : α → β) [Decida
   rw [← pos_iff_ne_zero, card_pos, fiber_nonempty_iff_mem_image]
 #align finset.fiber_card_ne_zero_iff_mem_image Finset.fiber_card_ne_zero_iff_mem_image
 
+lemma filter_card_le_iff (s : Finset α) (P : α → Prop) [DecidablePred P] (n : ℕ) :
+    (s.filter P).card ≤ n ↔ ∀ s' ≤ s, n < s'.card → ∃ a ∈ s', ¬ P a :=
+  (Multiset.filter_card_le_iff s.1 P n).trans ⟨fun H s' hs' h ↦ H s'.1 (by aesop) h,
+    fun H s' hs' h ↦ H ⟨s', nodup_of_le hs' s.2⟩ (fun x hx ↦ Multiset.subset_of_le hs' hx) h⟩
```
Comment (alreadydone) [inline Mathlib/Data/Finset/Card.lean:273]:
```suggestion
  (s.1.filter_card_le_iff P n).trans ⟨fun H s' hs' h ↦ H s'.1 (by aesop) h,
    fun H s' hs' h ↦ H ⟨s', nodup_of_le hs' s.2⟩ (fun x hx ↦ subset_of_le hs' hx) h⟩
```

**Review by alreadydone — APPROVED** [review]

**Review by alreadydone — APPROVED** [review]
maintainer merge

File: Mathlib/Data/Finset/Card.lean [hunk 3]
```diff
@@ -267,6 +267,11 @@ theorem fiber_card_ne_zero_iff_mem_image (s : Finset α) (f : α → β) [Decida
   rw [← pos_iff_ne_zero, card_pos, fiber_nonempty_iff_mem_image]
 #align finset.fiber_card_ne_zero_iff_mem_image Finset.fiber_card_ne_zero_iff_mem_image
 
+lemma filter_card_le_iff (s : Finset α) (P : α → Prop) [DecidablePred P] (n : ℕ) :
```
Comment (YaelDillies) [inline Mathlib/Data/Finset/Card.lean:270]:
```suggestion
lemma card_filter_le_iff (s : Finset α) (P : α → Prop) [DecidablePred P] (n : ℕ) :
```

File: Mathlib/Data/Finset/Card.lean [hunk 4]
```diff
@@ -267,6 +267,11 @@ theorem fiber_card_ne_zero_iff_mem_image (s : Finset α) (f : α → β) [Decida
...
+    (s.filter P).card ≤ n ↔ ∀ s' ≤ s, n < s'.card → ∃ a ∈ s', ¬ P a :=
```
Comment (YaelDillies) [inline Mathlib/Data/Finset/Card.lean:271]:
```suggestion
    (s.filter P).card ≤ n ↔ ∀ s' ⊆ s, n < s'.card → ∃ a ∈ s', ¬ P a :=
```

File: Mathlib/Data/Multiset/Basic.lean [hunk 5]
```diff
@@ -2154,6 +2154,16 @@ lemma map_filter' {f : α → β} (hf : Injective f) (s : Multiset α)
   simp [(· ∘ ·), map_filter, hf.eq_iff]
 #align multiset.map_filter' Multiset.map_filter'
 
+lemma filter_card_le_iff (s : Multiset α) (P : α → Prop) [DecidablePred P] (n : ℕ) :
```
Comment (YaelDillies) [inline Mathlib/Data/Multiset/Basic.lean:2157]:
```suggestion
lemma card_filter_le_iff (s : Multiset α) (P : α → Prop) [DecidablePred P] (n : ℕ) :
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**